tribe-ext-pdf-tickets/index.php

```diff
@@ -0,0 +1,892 @@
+	 * @var string
+	 */
+	public $pdf_ticket_meta_key = '_tribe_ext_pdf_tickets_unique_id';
+
+	/**
+	 * The query argument key for the Attendee ID.
+	 *
+	 * @var string
+	 */
+	public $pdf_unique_id_query_arg_key = 'tribe_ext_pdf_tickets_unique_id';
+
+	/**
+	 * The query argument key for retrying loading an attempted PDF.
+	 *
+	 * @var string
+	 */
+	public $pdf_retry_url_query_arg_key = 'tribe_ext_pdf_tickets_retry';
+
+	/**
+	 * An array of the absolute file paths of the PDF(s) to be attached
+	 * to the ticket email.
+	 *
+	 * One PDF attachment per attendee, even in a single order.
+	 *
+	 * @var array
+	 */
+	protected $attachments_array = array();
+
+	/**
+	 * Setup the Extension's properties.
+	 *
+	 * This always executes even if the required plugins are not present.
+	 */
+	public function construct() {
+		$this->add_required_plugin( 'Tribe__Tickets__Main', '4.5.2' );
+
+		add_action( 'tribe_plugins_loaded', array( $this, 'required_tribe_classes' ), 0 );
+
+		$this->set_url( 'https://theeventscalendar.com/extensions/pdf-tickets/' );
+
+		/**
+		 * When plugin is activated, add rewrite rules and then flush rewrite
+		 * rules. When plugin is deactivated, just flush rewrite rules.
+		 *
+		 * @link https://developer.wordpress.org/reference/functions/flush_rewrite_rules/#comment-597
+		 */
+		register_activation_hook( __FILE__, array( $this, 'flush_rewrite_rules_on_activation' ) );
+		register_deactivation_hook( __FILE__, array( $this, 'flush_rewrite_rules' ) );
+	}
+
+	/**
+	 * Check required plugins after all Tribe plugins have loaded.
+	 */
+	public function required_tribe_classes() {
+		if ( Tribe__Dependency::instance()->is_plugin_active( 'Tribe__Tickets_Plus__Main' ) ) {
+			$this->add_required_plugin( 'Tribe__Tickets_Plus__Main', '4.5.6' );
+
+			if ( Tribe__Dependency::instance()->is_plugin_active( 'Tribe__Events__Community__Tickets__Main' ) ) {
+				$this->add_required_plugin( 'Tribe__Events__Community__Tickets__Main', '4.4.3' );
+			}
+
+		}
+	}
+
+	/**
+	 * Extension initialization and hooks.
+	 *
+	 * mPDF version 7.0+ requires PHP 5.6+ with the mbstring and gd extensions.
+	 * Permalinks are required to be set in order to use this plugin. If they
+	 * are not set, display an informative admin error with a link to the
+	 * Permalink Settings admin screen and do not load the rest of this plugin.
+	 */
+	public function init() {
+		if ( version_compare( PHP_VERSION, '5.6', '<' ) ) {
+			$message = '<p>' . $this->get_name();
+
+			$message .= __( ' requires PHP 5.6 or newer to work (as well as the `mbstring` and `gd` PHP extensions). Please contact your website host and inquire about updating PHP.', 'tribe-extension' );
+
+			$message .= '</p>';
+
+			tribe_notice( $this->get_name(), $message, 'type=error' );
+
+			return;
+		}
+
+		if ( ! empty( get_option( 'permalink_structure' ) ) ) {
+			// Event Tickets
+			add_action( 'event_tickets_attendees_table_row_actions', array( $this, 'pdf_attendee_table_row_actions' ), 0, 2 );
+
```

Comment:
I won't call out each instance of this, but again I think the extra line breaks between `add_action()` calls are unnecessary and it would be nicer to remove and group together related calls, something like:

```php
// Event Tickets
add_action( 'ticket_hook_1', 'Some::callback_1' );
add_action( 'ticket_hook_2', 'Some::callback_2' );

// Event Tickets Plus
add_action( 'woo_specific', 'foo' );
add_action( 'edd_specific', 'bar' );
```

...and so on. That said, I'll leave the final decision up to you :-)